lib/scholar/neighbors/random_projection_forest.ex

```diff
@@ -0,0 +1,140 @@
+
+  def grow(tensor, num_trees, min_leaf_size) do
+    key = Nx.Random.key(System.system_time())
+    grow(tensor, num_trees, min_leaf_size, key)
+  end
+
+  def grow(tensor, num_trees, min_leaf_size, key) do
+    {size, dim} = Nx.shape(tensor)
+    {depth, leaf_size} = compute_depth_and_leaf_size(size, min_leaf_size, 0)
+    num_nodes = 2 ** depth - 1
+
+    {hyperplanes, _key} =
+      Nx.Random.normal(key, type: to_float_type(tensor), shape: {num_trees, depth, dim})
+
+    medians = Nx.broadcast(:nan, {num_trees, num_nodes})
+    # TODO: Maybe rename acc to indices
+    acc = Nx.broadcast(Nx.u32(0), {num_trees, size})
+    root = 0
+    start_index = 0
+    indices = Nx.iota({size}, type: :u32) |> Nx.new_axis(0) |> Nx.broadcast({num_trees, size})
+
+    {indices, medians} =
+      recur([{root, start_index, indices}], [], acc, leaf_size, 0, tensor, hyperplanes, medians)
+
+    %__MODULE__{
+      depth: depth,
+      leaf_size: leaf_size,
+      num_trees: num_trees,
+      indices: indices,
+      data: tensor,
+      hyperplanes: hyperplanes,
+      medians: medians
+    }
+  end
+
+  defp recur([], [], acc, _, _, _, _, medians) do
+    {acc, medians}
+  end
+
+  defp recur([], next, acc, leaf_size, level, tensor, hyperplanes, medians) do
+    recur(next, [], acc, leaf_size, level + 1, tensor, hyperplanes, medians)
+  end
+
+  defp recur(
+         [{node, start_index, indices} | rest],
+         next,
+         acc,
+         leaf_size,
+         level,
+         tensor,
+         hyperplanes,
+         medians
+       ) do
+    size = Nx.axis_size(indices, 1)
+
+    if size > leaf_size do
+      hyperplane = hyperplanes[[.., level]]
+
+      {median, left_indices, right_indices} =
+        Nx.Defn.jit_apply(&split(&1, &2, &3), [tensor, indices, hyperplane])
+
+      medians = Nx.put_slice(medians, [0, node], Nx.new_axis(median, 1))
+      child_start_index = start_index
+      next = [{left_child(node), child_start_index, left_indices} | next]
+      child_start_index = start_index + div(size, 2) + rem(size, 2)
+      next = [{right_child(node), child_start_index, right_indices} | next]
+      recur(rest, next, acc, leaf_size, level, tensor, hyperplanes, medians)
+    else
+      acc = Nx.put_slice(acc, [0, start_index], indices)
+      recur(rest, next, acc, leaf_size, level, tensor, hyperplanes, medians)
+    end
+  end
+
+  defnp split(tensor, indices, hyperplane) do
+    size = Nx.axis_size(indices, 1)
+    slice = Nx.take(tensor, indices)
+    proj = Nx.dot(slice, [2], [0], hyperplane, [1], [0])
+    sorted_indices = Nx.argsort(proj, axis: 1)
+    proj = Nx.take_along_axis(proj, sorted_indices, axis: 1)
+
+    median =
+      if rem(size, 2) == 1 do
+        proj[[.., div(size, 2)]]
+      else
+        mid = Nx.slice_along_axis(proj, div(size, 2) - 1, 2, axis: 1)
+        Nx.mean(mid, axes: [1])
+      end
+
+    indices = Nx.take_along_axis(indices, sorted_indices, axis: 1)
+    left_size = div(size, 2) + rem(size, 2)
```

Comment:
I would move div(size, 2) and rem(size, 2) to a separate variables since they are used multiple times